Confusing/invalid comment on valid Pull Requests

In https://github.com/carpentries/lesson-development-training/pull/165, the author is making rapid changes to several files at once that is faster than the actions are being triggered. Unfortunately, by the time, these workflows get to the validator, the HEAD has changed and it becomes invalid. 

Examples:

https://github.com/carpentries/lesson-development-training/commit/7655e136d566c0e9f4cd1004ae0963df63228de2 was made 50 seconds before
https://github.com/carpentries/lesson-development-training/commit/8f8f7c0bf74f2214f245db9ea1d1279464dd9a2a, but the runs did not start in that order: 

<details>
<summary> Log from 8f8f7c0bf74f2214f245db9ea1d1279464dd9a2a started on 2022-12-07T15:18:24.2470400Z</summary>

https://github.com/carpentries/lesson-development-training/actions/runs/3640242771/jobs/6144652715 

```
2022-12-07T15:18:24.2470400Z Requested labels: ubuntu-latest
2022-12-07T15:18:24.2470442Z Job defined at: carpentries/lesson-development-training/.github/workflows/pr-comment.yaml@refs/heads/main
2022-12-07T15:18:24.2470462Z Waiting for a runner to pick up this job...
2022-12-07T15:18:24.6613967Z Job is waiting for a hosted runner to come online.
2022-12-07T15:18:28.6411455Z Job is about to start running on the hosted runner: Hosted Agent (hosted)
2022-12-07T15:18:30.9683489Z Current runner version: '2.299.1'
2022-12-07T15:18:30.9710118Z ##[group]Operating System
2022-12-07T15:18:30.9710670Z Ubuntu
2022-12-07T15:18:30.9711001Z 22.04.1
2022-12-07T15:18:30.9711288Z LTS
2022-12-07T15:18:30.9711558Z ##[endgroup]
2022-12-07T15:18:30.9711887Z ##[group]Runner Image
2022-12-07T15:18:30.9712263Z Image: ubuntu-22.04
2022-12-07T15:18:30.9712622Z Version: 20221127.1
2022-12-07T15:18:30.9713097Z Included Software: https://github.com/actions/runner-images/blob/ubuntu22/20221127.1/images/linux/Ubuntu2204-Readme.md
2022-12-07T15:18:30.9713727Z Image Release: https://github.com/actions/runner-images/releases/tag/ubuntu22%2F20221127.1
2022-12-07T15:18:30.9714173Z ##[endgroup]
2022-12-07T15:18:30.9714472Z ##[group]Runner Image Provisioner
2022-12-07T15:18:30.9714846Z 2.0.91.1
2022-12-07T15:18:30.9715150Z ##[endgroup]
2022-12-07T15:18:30.9716076Z ##[group]GITHUB_TOKEN Permissions
2022-12-07T15:18:30.9716699Z Actions: write
2022-12-07T15:18:30.9717074Z Checks: write
2022-12-07T15:18:30.9717553Z Contents: write
2022-12-07T15:18:30.9717893Z Deployments: write
2022-12-07T15:18:30.9718258Z Discussions: write
2022-12-07T15:18:30.9718593Z Issues: write
2022-12-07T15:18:30.9718854Z Metadata: read
2022-12-07T15:18:30.9719175Z Packages: write
2022-12-07T15:18:30.9719500Z Pages: write
2022-12-07T15:18:30.9720126Z PullRequests: write
2022-12-07T15:18:30.9720541Z RepositoryProjects: write
2022-12-07T15:18:30.9720913Z SecurityEvents: write
2022-12-07T15:18:30.9721228Z Statuses: write
2022-12-07T15:18:30.9721552Z ##[endgroup]
2022-12-07T15:18:30.9725356Z Secret source: Actions
2022-12-07T15:18:30.9725818Z Prepare workflow directory
2022-12-07T15:18:31.0527335Z Prepare all required actions
2022-12-07T15:18:31.0711475Z Getting action download info
2022-12-07T15:18:31.4254030Z Download action repository 'carpentries/actions@main' (SHA:3b76ed566c6f52f8f15c6bf7718b9811ecf47ed1)
2022-12-07T15:18:32.8927488Z ##[group]Run carpentries/actions/download-workflow-artifact@main
2022-12-07T15:18:32.8927872Z with:
2022-12-07T15:18:32.8928093Z   run: 3640224573
2022-12-07T15:18:32.8928282Z   name: pr
2022-12-07T15:18:32.8928608Z   dir: /home/runner/work/lesson-development-training/lesson-development-training
2022-12-07T15:18:32.8928991Z   repo: carpentries/lesson-development-training
2022-12-07T15:18:32.8929465Z   token: ***
2022-12-07T15:18:32.8929679Z ##[endgroup]
2022-12-07T15:18:33.7670679Z ##[group]Run unzip pr.zip
2022-12-07T15:18:33.7671066Z [36;1munzip pr.zip[0m
2022-12-07T15:18:33.7671313Z [36;1mecho "NUM=$(<./NR)" >> $GITHUB_OUTPUT[0m
2022-12-07T15:18:33.7723874Z shell: /usr/bin/bash -e {0}
2022-12-07T15:18:33.7724152Z ##[endgroup]
2022-12-07T15:18:33.7945009Z Archive:  pr.zip
2022-12-07T15:18:33.7946378Z   inflating: NR                      
2022-12-07T15:18:33.7981277Z ##[group]Run echo "json<<EOF
2022-12-07T15:18:33.7981620Z [36;1mecho "json<<EOF[0m
2022-12-07T15:18:33.7982016Z [36;1m$(curl -sL https://files.carpentries.org/invalid-hashes.json)[0m
2022-12-07T15:18:33.7982332Z [36;1mEOF" >> $GITHUB_OUTPUT[0m
2022-12-07T15:18:33.8029265Z shell: /usr/bin/bash -e {0}
2022-12-07T15:18:33.8029559Z ##[endgroup]
2022-12-07T15:18:34.0667706Z ##[group]Run carpentries/actions/check-valid-pr@main
2022-12-07T15:18:34.0667993Z with:
2022-12-07T15:18:34.0668175Z   pr: 165
2022-12-07T15:18:34.0668408Z   sha: 8f8f7c0bf74f2214f245db9ea1d1279464dd9a2a
2022-12-07T15:18:34.0668638Z   fail_on_error: true
2022-12-07T15:18:34.0668899Z   repo: carpentries/lesson-development-training
2022-12-07T15:18:34.0669338Z   token: ***
2022-12-07T15:18:34.0669552Z ##[endgroup]
2022-12-07T15:18:34.8863915Z Files in PR: episodes/01-introduction.md, episodes/02-lesson-design.md, episodes/03-audience.md, episodes/05-objectives.md, episodes/13-exercises.md, episodes/15-narrative.md, episodes/17-explanation.md, episodes/19-operations.md, episodes/20-preparing.md, episodes/22-reflecting.md, episodes/24-external.md, episodes/26-collaborating.md
2022-12-07T15:18:34.8865652Z Is valid?: true
2022-12-07T15:18:34.9066862Z Evaluate and set job outputs
2022-12-07T15:18:34.9080946Z Set output 'is_valid'
2022-12-07T15:18:34.9107225Z Set output 'payload'
2022-12-07T15:18:34.9107514Z Set output 'number'
2022-12-07T15:18:34.9108010Z Set output 'msg'
2022-12-07T15:18:34.9108623Z Cleaning up orphan processes
```

</details>

<details>
<summary> Log from 7655e136d566c0e9f4cd1004ae0963df63228de2 started on 2022-12-07T15:19:07.3763427Z</summary>

https://github.com/carpentries/lesson-development-training/actions/runs/3640248674/jobs/6144666259

```
2022-12-07T15:19:07.3763427Z Requested labels: ubuntu-latest
2022-12-07T15:19:07.3763465Z Job defined at: carpentries/lesson-development-training/.github/workflows/pr-comment.yaml@refs/heads/main
2022-12-07T15:19:07.3763484Z Waiting for a runner to pick up this job...
2022-12-07T15:19:07.5748519Z Job is waiting for a hosted runner to come online.
2022-12-07T15:19:11.2792444Z Job is about to start running on the hosted runner: GitHub Actions 2 (hosted)
2022-12-07T15:19:14.1868616Z Current runner version: '2.299.1'
2022-12-07T15:19:14.1898604Z ##[group]Operating System
2022-12-07T15:19:14.1899367Z Ubuntu
2022-12-07T15:19:14.1899659Z 22.04.1
2022-12-07T15:19:14.1900018Z LTS
2022-12-07T15:19:14.1900380Z ##[endgroup]
2022-12-07T15:19:14.1900703Z ##[group]Runner Image
2022-12-07T15:19:14.1901155Z Image: ubuntu-22.04
2022-12-07T15:19:14.1901535Z Version: 20221204.2
2022-12-07T15:19:14.1902362Z Included Software: https://github.com/actions/runner-images/blob/ubuntu22/20221204.2/images/linux/Ubuntu2204-Readme.md
2022-12-07T15:19:14.1903124Z Image Release: https://github.com/actions/runner-images/releases/tag/ubuntu22%2F20221204.2
2022-12-07T15:19:14.1903680Z ##[endgroup]
2022-12-07T15:19:14.1906796Z ##[group]Runner Image Provisioner
2022-12-07T15:19:14.1907167Z 2.0.91.1
2022-12-07T15:19:14.1907531Z ##[endgroup]
2022-12-07T15:19:14.1908982Z ##[group]GITHUB_TOKEN Permissions
2022-12-07T15:19:14.1909909Z Actions: write
2022-12-07T15:19:14.1910246Z Checks: write
2022-12-07T15:19:14.1910851Z Contents: write
2022-12-07T15:19:14.1911336Z Deployments: write
2022-12-07T15:19:14.1911678Z Discussions: write
2022-12-07T15:19:14.1912090Z Issues: write
2022-12-07T15:19:14.1912458Z Metadata: read
2022-12-07T15:19:14.1912779Z Packages: write
2022-12-07T15:19:14.1913182Z Pages: write
2022-12-07T15:19:14.1913618Z PullRequests: write
2022-12-07T15:19:14.1913988Z RepositoryProjects: write
2022-12-07T15:19:14.1914417Z SecurityEvents: write
2022-12-07T15:19:14.1914844Z Statuses: write
2022-12-07T15:19:14.1915166Z ##[endgroup]
2022-12-07T15:19:14.1919386Z Secret source: Actions
2022-12-07T15:19:14.1919968Z Prepare workflow directory
2022-12-07T15:19:14.2893285Z Prepare all required actions
2022-12-07T15:19:14.3138385Z Getting action download info
2022-12-07T15:19:14.6130131Z Download action repository 'carpentries/actions@main' (SHA:3b76ed566c6f52f8f15c6bf7718b9811ecf47ed1)
2022-12-07T15:19:16.2940113Z ##[group]Run carpentries/actions/download-workflow-artifact@main
2022-12-07T15:19:16.2940537Z with:
2022-12-07T15:19:16.2940799Z   run: 3640217742
2022-12-07T15:19:16.2941033Z   name: pr
2022-12-07T15:19:16.2941416Z   dir: /home/runner/work/lesson-development-training/lesson-development-training
2022-12-07T15:19:16.2941858Z   repo: carpentries/lesson-development-training
2022-12-07T15:19:16.2942414Z   token: ***
2022-12-07T15:19:16.2942662Z ##[endgroup]
2022-12-07T15:19:17.4126213Z ##[group]Run unzip pr.zip
2022-12-07T15:19:17.4126625Z [36;1munzip pr.zip[0m
2022-12-07T15:19:17.4126902Z [36;1mecho "NUM=$(<./NR)" >> $GITHUB_OUTPUT[0m
2022-12-07T15:19:17.4189048Z shell: /usr/bin/bash -e {0}
2022-12-07T15:19:17.4189405Z ##[endgroup]
2022-12-07T15:19:17.4448699Z Archive:  pr.zip
2022-12-07T15:19:17.4502494Z   inflating: NR                      
2022-12-07T15:19:17.4565350Z ##[group]Run echo "json<<EOF
2022-12-07T15:19:17.4565747Z [36;1mecho "json<<EOF[0m
2022-12-07T15:19:17.4566259Z [36;1m$(curl -sL https://files.carpentries.org/invalid-hashes.json)[0m
2022-12-07T15:19:17.4566679Z [36;1mEOF" >> $GITHUB_OUTPUT[0m
2022-12-07T15:19:17.4629569Z shell: /usr/bin/bash -e {0}
2022-12-07T15:19:17.4629939Z ##[endgroup]
2022-12-07T15:19:18.0585199Z ##[group]Run carpentries/actions/check-valid-pr@main
2022-12-07T15:19:18.0585558Z with:
2022-12-07T15:19:18.0585784Z   pr: 165
2022-12-07T15:19:18.0586057Z   sha: 7655e136d566c0e9f4cd1004ae0963df63228de2
2022-12-07T15:19:18.0586330Z   fail_on_error: true
2022-12-07T15:19:18.0586647Z   repo: carpentries/lesson-development-training
2022-12-07T15:19:18.0587186Z   token: ***
2022-12-07T15:19:18.0587424Z ##[endgroup]
2022-12-07T15:19:18.6744744Z ##[error]
## :x: DANGER :x: 

**Do not merge this Pull Request**. This PR has been spoofed to look like #165 (HEAD @ 8f8f7c0bf74f2214f245db9ea1d1279464dd9a2a).
2022-12-07T15:19:18.6941707Z Evaluate and set job outputs
2022-12-07T15:19:18.6958244Z Set output 'is_valid'
2022-12-07T15:19:18.6959262Z Set output 'number'
2022-12-07T15:19:18.6959857Z Set output 'msg'
2022-12-07T15:19:18.6961757Z Cleaning up orphan processes
```

</details>


I do not yet know the solution to this, but using `concurrency` on the pr receiving workflows may help prevent this from happening: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#concurrency

--- a/DESCRIPTION
+++ b/DESCRIPTION
@@ -1,6 +1,6 @@
 Package: sandpaper
 Title: Create and Curate Carpentries Lessons
-Version: 0.11.2
+Version: 0.11.3
 Authors@R: c(
     person(given = "Zhian N.",
            family = "Kamvar",

--- a/NEWS.md
+++ b/NEWS.md
@@ -1,3 +1,16 @@
+# sandpaper 0.11.3
+
+CONTINUOUS INTEGRATION
+----------------------
+
+* Pull Request workflows will now automatically cancel if several commits are
+  sent in succession. Specifically, the workflows `pr-recieve.yaml` and 
+  `pr-comment.yaml` are given separate `concurrency` parameters based on the
+  branch name and the pull request number. These concurrencies will prevent
+  false alarms as found in 
+  <https://github.com/carpentries/lesson-development-training/pull/165#issuecomment-1337182275>.
+  (discovered: @anenadic; reported: @zkamvar, #374; fixed: @zkamvar, #376)
+
 # sandpaper 0.11.2
 
 MISC

--- a/inst/workflows/README.md
+++ b/inst/workflows/README.md
@@ -147,6 +147,11 @@ pull request. GitHub has safeguarded the token used in this workflow to have no
 priviledges in the repository, but we have taken precautions to protect against
 spoofing.
 
+This workflow is triggered with every push to a pull request. If this workflow
+is already running and a new push is sent to the pull request, the workflow
+running from the previous push will be cancelled and a new workflow run will be
+started.
+
 The first step of this workflow is to check if it is valid (e.g. that no
 workflow files have been modified). If there are workflow files that have been
 modified, a comment is made that indicates that the workflow is not run. If 
@@ -160,7 +165,7 @@ request. This builds the content and uploads three artifacts:
 3. The rendered files (build)
 
 Because this workflow builds generated content, it follows the same general 
-process as the sandpaper-main workflow with the same caching mechanisms.
+process as the `sandpaper-main` workflow with the same caching mechanisms.
 
 The artifacts produced are used by the next workflow.
 
@@ -176,7 +181,7 @@ The steps in this workflow are:
 3. If it is valid: update the pull request comment with the summary of changes
 
 Importantly: if the pull request is invalid, the branch is not created so any
-malicious code is not published. 
+malicious code is not published.
 
 From here, the maintainer can request changes from the author and eventually 
 either merge or reject the PR. When this happens, if the PR was valid, the 

--- a/inst/workflows/pr-comment.yaml
+++ b/inst/workflows/pr-comment.yaml
@@ -63,6 +63,7 @@ jobs:
         with:
           pr: ${{ steps.get-pr.outputs.NUM }}
           sha: ${{ github.event.workflow_run.head_sha }}
+          headroom: 3 # if it's within the last three commits, we can keep going, because it's likely rapid-fire
           invalid: ${{ fromJSON(steps.hash.outputs.json)[github.repository] }}
           fail_on_error: true
 

--- a/inst/workflows/pr-comment.yaml
+++ b/inst/workflows/pr-comment.yaml
@@ -8,6 +8,11 @@ on:
     types:
       - completed
 
+concurrency:
+  group: pr-${{ github.event.workflow_run.pull_requests[0].number }}
+  cancel-in-progress: true
+
+
 jobs:
   # Pull requests are valid if:
   #  - they match the sha of the workflow run head commit

--- a/inst/workflows/pr-receive.yaml
+++ b/inst/workflows/pr-receive.yaml
@@ -5,6 +5,10 @@ on:
     types:
       [opened, synchronize, reopened]
 
+concurrency:
+  group: ${{ github.ref }}
+  cancel-in-progress: true
+
 jobs:
   test-pr:
     name: "Record PR number"

